src/java/org/apache/cassandra/index/sai/SSTableContext.java

```diff
@@ -17,274 +17,57 @@
  */
 package org.apache.cassandra.index.sai;
 
-import java.io.IOException;
-
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
-
-import org.apache.cassandra.db.DecoratedKey;
-import org.apache.cassandra.index.sai.disk.io.IndexComponents;
-import org.apache.cassandra.index.sai.disk.v1.BlockPackedReader;
-import org.apache.cassandra.index.sai.disk.v1.MetadataSource;
-import org.apache.cassandra.index.sai.disk.v1.MonotonicBlockPackedReader;
-import org.apache.cassandra.index.sai.utils.LongArray;
-import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.V1SSTableContext;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
 import org.apache.cassandra.io.sstable.format.SSTableReader;
-import org.apache.cassandra.io.util.FileHandle;
-import org.apache.cassandra.io.util.RandomAccessReader;
-import org.apache.cassandra.utils.Throwables;
-import org.apache.cassandra.utils.concurrent.Ref;
 import org.apache.cassandra.utils.concurrent.RefCounted;
 import org.apache.cassandra.utils.concurrent.SharedCloseableImpl;
 
-import static org.apache.cassandra.index.sai.disk.OnDiskKeyProducer.NO_OFFSET;
-
 /**
  * SSTableContext is created for individual sstable shared across indexes to track per-sstable index files.
  *
  * SSTableContext itself will be released when receiving sstable removed notification, but its shared copies in individual
  * SSTableIndex will be released when in-flight read requests complete.
  */
-public class SSTableContext extends SharedCloseableImpl
+public abstract class SSTableContext extends SharedCloseableImpl
```

Comment:
Why do we need multiple versions of it? It would be nice to have all the format-specific things in the `OnDiskFormat` (or at least delegated through `OnDiskFormat`).

Reply:
Yes, I agree. The only issue I have is finding a version specific abstraction to hold to per-sstable files readers. These are currently held in the `SSTableContext` so they can be shared amongst the segment readers. The `Segment` and it's related classes are already in `v1` so I'll see if I can push all the file handling there as well.

I need to move the query path into the versioning code so it is likely I can do this as part of that.